spock-core/src/main/java/org/spockframework/util/JavaProcessThreadDumpCollector.java

```diff
@@ -0,0 +1,115 @@
+
+  static JavaProcessThreadDumpCollector create(ThreadDumpUtilityType utility) {
+    try {
+      return new FunctionalJavaProcessThreadDumpCollector(utility);
+    } catch (Exception e) {
+      ByteArrayOutputStream stream = new ByteArrayOutputStream();
+      e.printStackTrace(new PrintStream(stream));
+      System.out.printf("Thread dump capturing is not available: " + stream);
+      return NO_OP;
+    }
+  }
+
+  JavaProcessThreadDumpCollector NO_OP = __ -> {
+  };
+
+  class FunctionalJavaProcessThreadDumpCollector implements JavaProcessThreadDumpCollector {
+
+    private static final String JAVA_HOME_SYS_PROP = "java.home";
+
+    private final ThreadDumpUtilityType utility;
+    private final List<String> command;
+
+    public FunctionalJavaProcessThreadDumpCollector(ThreadDumpUtilityType utility) {
+      long pid = currentProcessId();
+      Path utilityPath = getUtilityCommandPath(utility);
+
+      this.utility = utility;
+      this.command = utility == ThreadDumpUtilityType.JSTACK
+        ? Arrays.asList(utilityPath.toString(), Long.toString(pid))
+        : Arrays.asList(utilityPath.toString(), Long.toString(pid), "Thread.print");
+    }
+
+    @Override
+    public void appendThreadDumpOfCurrentJvm(StringBuilder builder) throws IOException, InterruptedException {
+      builder.append("Thread dump of current JVM (")
+        .append(utility.name())
+        .append("):\n");
+      builder.append("------------------------------")
+        .append(TextUtil.repeatChar('-', utility.name().length()))
+        .append("\n");
+
+      Process process = new ProcessBuilder(command)
+        .redirectErrorStream(true)
+        .start();
+
+      captureProcessOutput(process, builder);
+      process.waitFor();
+    }
+
+    private void captureProcessOutput(Process process, StringBuilder builder) throws IOException {
+      try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+        String line;
+        while ((line = reader.readLine()) != null) {
+          builder.append(line).append("\n");
+        }
+      }
+    }
+
+    private static Path getUtilityCommandPath(ThreadDumpUtilityType utility) {
+      Path utilityPath = utility.getPath(getJavaHome());
+      if (!Files.exists(utilityPath)) {
+        throw new SpockException("Could not find requested thread dump capturing utility '" + utility.name().toLowerCase() + "' under expected path '" + utilityPath + "'");
+      }
+      return utilityPath;
+    }
+
+    private static Path getJavaHome() {
+      return Optional.ofNullable(System.getProperty(JAVA_HOME_SYS_PROP))
+        .map(Paths::get)
+        .orElseThrow(() -> new SpockException("Could not determine java home directory, as 'java.home' system property is not set"));
+    }
+
+    private static long currentProcessId() {
+      try {
+        String jvmName = ManagementFactory.getRuntimeMXBean().getName();
+        return Long.parseLong(jvmName.split("@")[0]);
```

Comment:
💬 It is not guaranteed to work, as the format might be completely different, but in most cases, this should do the trick.